### PR TITLE
ELPP-3205 Emit notifications when Profiles change

### PR DESCRIPTION
**Pull request by seanwiseman**
I have included functionality to configure, initialise and publish via `elife_bus_sdk`.

I need assistance with making sure that the db commit(s) are being executed in the right place, due to `SQLALCHEMY_COMMIT_ON_TEARDOWN`now being set to `False` to allow confirmation that a transaction has been successful before publishing an event to the bus. All the tests are passing after I mocked out the publisher, but there might be things that aren't being captured by the current tests now that `SQLALCHEMY_COMMIT_ON_TEARDOWN` is off.

### CODE REVIEW

File: app.cfg.dist [hunk 1]
```diff
@@ -12,3 +12,9 @@ authorize_uri = https://sandbox.orcid.org/oauth/authorize
 token_uri = https://sandbox.orcid.org/oauth/token
 client_id = client_id
 client_secret = client_secret
+
+[bus]
```
Comment (giorgiosironi) [inline app.cfg.dist]: https://github.com/elifesciences/profiles-formula/blob/master/salt/profiles/config/srv-profiles-app.cfg needs updating, using some pillar values; `{{ pillar.elife.env}}` is good for `env`, `name` can be hardcoded, the other two need configuring.
These are the existing topics:
```
        {
            "TopicArn": "arn:aws:sns:us-east-1:512686554592:bus-profiles--ci"
        }, 
        {
            "TopicArn": "arn:aws:sns:us-east-1:512686554592:bus-profiles--continuumtest"
        }, 
        {
            "TopicArn": "arn:aws:sns:us-east-1:512686554592:bus-profiles--end2end"
        }, 
        {
            "TopicArn": "arn:aws:sns:us-east-1:512686554592:bus-profiles--prod"
        }, 
```

File: manage.py [hunk 2]
```diff
@@ -21,6 +22,15 @@
 
 CONFIG = create_app_config(CONFIG_FILE)
 configure_logging(CONFIG.name, getattr(logging, CONFIG.logging['level']), CONFIG.logging['path'])
+
+PUBLISHER = get_publisher(pub_name='profiles',
+                          config={
+                              'region': CONFIG.bus['region'],
+                              'subscirber': CONFIG.bus['subscriber'],
```
Comment (giorgiosironi) [inline manage.py]: typo?

File: profiles/api/oauth2.py [hunk 3]
```diff
@@ -21,6 +24,7 @@
 from profiles.utilities import expires_at, remove_none_values
 
 LOGGER = logging.getLogger(__name__)
+PUBLISHER = get_publisher('profiles')
 
 
 def create_blueprint(orcid: Dict[str, str], clients: Clients, profiles: Profiles,
```
Comment (giorgiosironi) [inline profiles/api/oauth2.py]: I think we should inject the publisher as an argument to `create_blueprint` to avoid global state

File: profiles/api/oauth2.py [hunk 4]
```diff
@@ -137,6 +141,12 @@ def _token() -> Response:
 
         _update_profile(profile, orcid_token)
 
+        # manually commit session here?
+        db.session.commit()
```
Comment (giorgiosironi) [inline profiles/api/oauth2.py]: perhaps this is good enough for now, it's a pain point to deal with sooner or later

File: profiles/config.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 class Config(ABC):
     name = 'unknown'
     DEBUG = False
-    SQLALCHEMY_COMMIT_ON_TEARDOWN = True
+    SQLALCHEMY_COMMIT_ON_TEARDOWN = False
```
Comment (giorgiosironi) [inline profiles/config.py]: However if this is turned to `False` there are probably many places where we need to commit the transaction manually.
The pattern is the [Unit Of Work](https://martinfowler.com/eaaCatalog/unitOfWork.html), SQLAlchemy implements it with `db.session`, but perhaps we can automate the calls to that with some decorator (magic word) to put around the routes?

File: manage.py (hunk 2 above)
Comment (seanwiseman) [reply manage.py]: 👍🏻

File: profiles/api/oauth2.py (hunk 3 above)
Comment (seanwiseman) [reply profiles/api/oauth2.py]: The `publisher` is now being injected as an argument into the `create_blueprint` function.

File: app.cfg.dist (hunk 1 above)
Comment (seanwiseman) [reply app.cfg.dist]: I have opened PRs for both Profiles-formula and Builder-private including the new values.

File: profiles/factory.py [hunk 6]
```diff
@@ -24,11 +25,14 @@ def create_app(config: Config, clients: Clients) -> Flask:
     orcid_client = OrcidClient(config.orcid['api_uri'])
     orcid_tokens = SQLAlchemyOrcidTokens(db)
     profiles = SQLAlchemyProfiles(db)
+
+    publisher = get_publisher(pub_name='profiles', config={**config.bus})
```
Comment (thewilkybarkid) [inline profiles/factory.py]: `**config.bus` is a bit magic, could be explicit and remove `env` (since it's duplicating `profiles.environment`).

File: .pylintrc [hunk 7]
```diff
@@ -1,5 +1,5 @@
 [MESSAGE CONTROL]
-disable = abstract-method,missing-docstring,redefined-outer-name,too-few-public-methods
+disable = abstract-method,missing-docstring,redefined-outer-name,too-few-public-methods, too-many-arguments
```
Comment (thewilkybarkid) [inline .pylintrc]: I guess that's caused by `oauth2.create_blueprint()`? Is there are better way to do DI?

Comment (seanwiseman) [reply .pylintrc]: Yes it was. You could wrap some of them up but the `create_blueprint` is creating the blueprint with x3 view functions and x3 utility functions so its having to accept args in turn for them all. If it was a 'normal' non factory function I would say reduce it, but I don't think its unacceptable to have 6 arguments for a function doing this much.

That said I normally follow fat models, thin views, so any business logic that deals with finding or updating a model stays in the model class (or possibly the `repository` in this case). That would leave the views a lot thinner and therefore probably require less arguments to initialise.

File: profiles/factory.py (hunk 6 above)
Comment (seanwiseman) [reply profiles/factory.py]: It was left to unpack due to the fact you can pass in additional arguments such as overriding `endpoint_url` and what not for the boto resource. (previous incarnations of this functionality included this so I did not want to get rid of it) 
```
# SNSPublisher
def __init__(self, region: str = '', subscriber: str = '', name: str = '', env: str = '', **overrides):
```
Though in `Profiles` at least its probably not going to need anything other than the values to build the Arn, so I don't mind being explicit. Will change. Will also using existing `profiles.env` value.

**Comment by seanwiseman** [comment]
I have now moved the send event logic into a function that will be executed each time model(s) are inserted, updated or removed. This is done by hooking into SQLAlchemy signals so I have been able to remove any code from the views as its not needed there, and have reverted `SQLALCHEMY_COMMIT_ON_TEARDOWN` back to `True` as (for this case at least) there is not a need for me to manually commit the session.

File: .pylintrc (hunk 7 above)
Comment (thewilkybarkid) [reply .pylintrc]: Can be reverted now? (We might hit it later though.)

File: profiles/events.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+from profiles.models import Profile
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def send_update_events(publisher: EventPublisher) -> Callable[..., None]:
+    def wrapper(sender: Flask, changes: tuple) -> None:  # pylint:disable=unused-argument
+        for instance, operation in changes:
+            if isinstance(instance, Profile):
```
Comment (thewilkybarkid) [inline profiles/events.py]: Does this cater for something like just an `Affiliation` change? We could do with tests here.

File: .pylintrc (hunk 7 above)
Comment (seanwiseman) [reply .pylintrc]: Yes that can go now. Will remove.

File: profiles/factory.py [hunk 9]
```diff
@@ -39,4 +47,6 @@ def create_app(config: Config, clients: Clients) -> Flask:
     app.register_error_handler(ClientError, errors.client_error_handler)
     app.register_error_handler(OAuth2Error, errors.oauth2_error_handler)
 
+    models_committed.connect(send_update_events(publisher=publisher), app)
```
Comment (thewilkybarkid) [inline profiles/factory.py]: That's nice and easy to use.

File: profiles/events.py (hunk 8 above)
Comment (seanwiseman) [reply profiles/events.py]: Currently I only act if the instance being saved is of type `Profile` though all model changes will pass through here so can be handled. Shall I add support for `Affiliation` and some tests surrounding it?

Comment (thewilkybarkid) [reply profiles/events.py]: It needs to be if _any_ part of a `Profile` changes (so `EmailAddress` too; what happens with composite types?)

Comment (seanwiseman) [reply profiles/events.py]: OK so `Profile`, `Affiliation` and `EmailAddress` will need to have slightly different handling but the composite type updates effectively count as field updates on the parent model they are attached to. So an `Address` thats changed and saved will trigger the associated `Affiliation` to be updated. I will add them and run the additional tests.

Comment (seanwiseman) [reply profiles/events.py]: OK so we now have support for updates of all models excluding `OrcidToken`.

I had refactor the event publisher function to keep a list of ids for a single `session.commit` and then filter out duplicates so you don't receive multiple messages for single events, for example:

- Updating a `Profile` directly would send one change event` [(<Profile test>, 'update')]` this is fine, one event is sent to SNS.

- Creating and adding an `Affiliation` would send two change events `[(<Affiliation 1>, 'insert'), (<Profile test>, 'update')]` in this instance only one event is sent to SNS because I filter out duplicate related profile ids.

- Updating an `Address` (or any composite types) would only send one change event (linked to the parent model) `[(<Affiliation 1>), 'update']` this is fine and one event will be sent to SNS.

In short, any sort of change or collection of changes to a single `Profile` will send only one event to SNS (including composite types).

File: profiles/events.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def send_update_events(publisher: EventPublisher) -> Callable[..., None]:
+    def wrapper(sender: Flask, changes: tuple) -> None:  # pylint:disable=unused-argument
+        ids = []
+
+        for instance, operation in changes:  # pylint:disable=unused-variable
+            if isinstance(instance, Profile):
+                ids.append(instance.id)
+            if isinstance(instance, (Affiliation, EmailAddress)):
```
Comment (giorgiosironi) [inline profiles/events.py]: It's a bit worrying that we would have to traverse the graphs from any modified object to go back to what is the `profile` being modified, but it's good for now :+1: 
It comes from the fact that the ORM is fine-grained for the single entities, but we are after a model where `1 transaction` == `1 Profile Aggregate` == `1 Profile object and any of the Value Object it contains`

Comment (seanwiseman) [reply profiles/events.py]: I agree. If it was not for composite types being used we could just look for `Profile` model changes as any direct change to any of its Foreign keys is picked up, its just that the Foreign key's composite types do not get classed as a `Profile` model update so you have to account of the other `db.models`.

But yes its fine for now.

Comment (thewilkybarkid) [reply profiles/events.py]: Composites are the problem? I'm not entirely happy with how they work, so we could just replace them with plain value objects. (Though this actually sounds like an upstream bug.)

Comment (seanwiseman) [reply profiles/events.py]: They are not a problem, they are just the current reason for having to check per model type. 

I am used to them being plain value objects, but there will still be an element of checking based on model types even if you do that. 

If say `Address` was a `db.Model` and had a foreign key relationship to an `Affiliation` and you updated the `Address` object, you would still only get updates up to the `Affiliation` (would not bubble up to a `Profile` update), that's why we have to pull the profile id from the `Affiliation` object in that case.

For this use case I think its fine, its only run on a `session.commit` and will only send one event message per profile update.

Comment (thewilkybarkid) [reply profiles/events.py]: Ah ok, I misunderstood.

**Comment by thewilkybarkid** [comment]
Looks good, I think we just need two more tests: one if a profile gets deleted, and an integration test (ie `test_oauth2`).

**Comment by seanwiseman** [comment]
I have added a test for profile deletion, I also made sure that the object information is still passed to the event (so we still have an `id` even if its just been deleted) and it is.

As for an integration test, really its just testing that SQLAlchemy actually delivers model event information when it says it should, its not our code per say as we already test how it handles these events when it receives them. Any how, I am getting into mock spaghetti at the moment trying to track the calls to the 'actual' callback function that was passed to the listener in the app in `create_app`. 
Will keep trying.

**Comment by giorgiosironi** [comment]
For what it's worth we will also test events integration in end2end, but there is no one listening yet where we could make the assertion.

**Comment by seanwiseman** [comment]
After further investigation its seems that event signals (handled by the `blinker` library) are not actually triggered when using app and db test fixtures. I even tried `SQLAlchemy` base events and they don't seem to fire in test cases either. When being used outside of the test environment they fire fine, its something to with how the fixtures are setup and torn down.

I have added a test to prove we have a handler registered on app init and then other tests to show how these events are handled, I just can't mock the 'sending' of the signals using the test db fixture at present. I think the end to end tests will cover what we need.

**Comment by giorgiosironi** [comment]
They will, as soon as we have a client for these events to check.
An intermediate level is also to perform integration tests against the real nginx+uwsgi service running on `http://localhost:80`. Not sure we need such a layer here, but it's an option.

**Comment by seanwiseman** [comment]
OK understood.
